src/hypothesis/strategies.py

```diff
@@ -412,13 +413,17 @@ def sampled_from(elements):
 
     from hypothesis.searchstrategy.misc import SampledFromStrategy, \
         JustStrategy
-    elements = tuple(iter(elements))
```

Comment:
Because you’ve removed this line, the `len(elements)` call further down may throw a `TypeError`. Consider the following example:

```python
from hypothesis import given
from hypothesis.strategies import sampled_from

def f():
    yield 1

@given(sampled_from(f()))
def test(x):
    pass
```

Old versions of Hypothesis would run this fine. We want to deprecate it because the iteration order (or output!) of generators isn’t fixed, but what actually happens is you get a `TypeError` on line 419 instead.

Casting to a tuple isn’t perfect either, because then I pass in an infinite iterator and everything explodes. How about changing 419 to read:

```python
try:
    if len(elements) == 1:
        return JustStrategy(elements[0])
except TypeError:
    pass
```